File: Languages/Haskell/BeeMovie.hs
module Main where

import System.IO
import Data.List
import Data.Char
import Data.Ord

-- Hello, World!
hello :: IO ()
hello = putStrLn "Hello, World!"

data LetterCount = LetterCount Char Int
    deriving (Show, Eq)

analyzeBeeMovie :: IO ()
analyzeBeeMovie = do
    let scriptPath = "../beemoviescript.txt"
    
    fileExists <- doesFileExist scriptPath
    if not fileExists
        then putStrLn $ "File not found: " ++ scriptPath
        else do
            putStrLn "Bee Movie Script:"
            putStrLn "--------------------------------------------------"
            
            content <- readFile scriptPath
            let lines' = lines content
            
            -- Print each line
            mapM_ putStrLn lines'
            
            -- Count letters
            let allChars = concat lines'
            let letters = filter isLetter allChars
            let lowerLetters = map toLower letters
            let letterCounts = countOccurrences lowerLetters
            
            putStrLn "--------------------------------------------------"
            putStrLn "Analysis complete."
            
            if not (null letters)
                then do
                    let sortedCounts = sortBy (comparing (\(LetterCount _ c) -> Down c)) letterCounts
                    putStrLn "\nTop 3 most commonly used letters:"
                    mapM_ printTop3 (take 3 sortedCounts)
                else putStrLn "No letters found in the script."

countOccurrences :: String -> [LetterCount]
countOccurrences str = 
    let uniqueLetters = nub $ sort str
    in map (\letter -> LetterCount letter (length $ filter (== letter) str)) uniqueLetters

printTop3 :: LetterCount -> IO ()
printTop3 (LetterCount letter count) = 
    putStrLn $ show letter ++ ": " ++ show count ++ " times"

main :: IO ()
main = do
    hello
    analyzeBeeMovie

File: Languages/Julia/bee_movie.jl
# Hello, World!
println("Hello, World!")

function analyze_bee_movie()
    script_path = "../beemoviescript.txt"
    
    if !isfile(script_path)
        println("File not found: $script_path")
        return
    end
    
    println("Bee Movie Script:")
    println("--------------------------------------------------")
    
    # Read file line by line and print each line
    lines = readlines(script_path)
    
    # Print each line
    for line in lines
        println(line)
    end
    
    # Count letters
    all_chars = join(lines, "")
    letters = filter(isletter, all_chars)
    lower_letters = lowercase.(letters)
    
    println("--------------------------------------------------")
    println("Analysis complete.")
    
    if length(letters) > 0
        letter_counts = countmap(lower_letters)
        sorted_counts = sort(collect(letter_counts), by = x -> x[2], rev = true)
        top_3 = first(sorted_counts, 3)
        
        println("\nTop 3 most commonly used letters:")
        for (i, (letter, count)) in enumerate(top_3)
            println("$i. '$letter': $count times")
        end
    else
        println("No letters found in the script.")
    end
end

analyze_bee_movie()
